### PR TITLE
Fix SimpleImputer error which occurs when all features are bool type

### LINKED ISSUE #1166: Boolean column to Imputer breaks if it is the only categorical column
If only a boolean column is passed to a SimpleImputer, or if a boolean column is the only non-numeric column passed to an Imputer, it breaks:

Repro: the following code fails with a `ValueError: SimpleImputer does not support data with dtype bool. Please provide either a numeric array (with a floating point or integer dtype) or categorical data represented either as an array with integer dtype or an array of string values with an object dtype.` check.
```
X = pd.DataFrame({
        "bool col with nan": pd.Series([True, np.nan, False, np.nan, True], dtype='bool'),
    })
imputer = Imputer()
imputer.fit(X, y)
```

However, as soon as we add another column (object, category, with or without np.nan), the imputer works fine:
```
X = pd.DataFrame({
        "object with nan": ["b", "b", np.nan, "c", np.nan],
        "bool col with nan": pd.Series([True, np.nan, False, np.nan, True], dtype='bool'),
    })
imputer = Imputer()
imputer.fit(X, y)
```

The stack trace is as follows:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-36-10afc6697657> in <module>
     23     })
     24 imputer = Imputer()
---> 25 imputer.fit(X, y)
     26 # lp.fit(X, y)

~/Desktop/evalml/evalml/utils/base_meta.py in _set_fit(self, X, y)
     12         @wraps(method)
     13         def _set_fit(self, X, y=None):
---> 14             return_value = method(self, X, y)
     15             self._is_fitted = True
     16             return return_value

~/Desktop/evalml/evalml/pipelines/components/transformers/imputers/imputer.py in fit(self, X, y)
     78         if len(X_categorical.columns) > 0:
     79             import pdb; pdb.set_trace()
---> 80             self._categorical_imputer.fit(X_categorical, y)
     81             self._categorical_cols = X_categorical.columns
     82         return self

~/Desktop/evalml/evalml/utils/base_meta.py in _set_fit(self, X, y)
     12         @wraps(method)
     13         def _set_fit(self, X, y=None):
---> 14             return_value = method(self, X, y)
     15             self._is_fitted = True
     16             return return_value

~/Desktop/evalml/evalml/pipelines/components/transformers/imputers/simple_imputer.py in fit(self, X, y)
     47         X = X.fillna(value=np.nan)
     48 
---> 49         self._component_obj.fit(X, y)
     50         self._all_null_cols = set(X.columns) - set(X.dropna(axis=1, how='all').columns)
     51         return self

~/Desktop/venv/lib/python3.7/site-packages/sklearn/impute/_base.py in fit(self, X, y)
    266         self : SimpleImputer
    267         """
--> 268         X = self._validate_input(X)
    269         super()._fit_indicator(X)
    270 

~/Desktop/venv/lib/python3.7/site-packages/sklearn/impute/_base.py in _validate_input(self, X)
    249                              "categorical data represented either as an array "
    250                              "with integer dtype or an array of string values "
--> 251                              "with an object dtype.".format(X.dtype))
    252 
    253         return X

ValueError: SimpleImputer does not support data with dtype bool. Please provide either a numeric array (with a floating point or integer dtype) or categorical data represented either as an array with integer dtype or an array of string values with an object dtype.
```

Not quite sure what we can do, given this looks like a check on the scikit-learn side (via `_validate_input`); we could pass in a fake column to appease this check?

### PR DESCRIPTION
**Pull request by christopherbunn**
Fixed issues with all-bool inputs for `Imputer` and `SimpleImputer` by internally converting to the category datatype and reconverting to bools at the end. 

Resolves #1166

### CODE REVIEW

**Review by jeremyliweishih — APPROVED** [review]
cool looks good!

File: evalml/tests/component_tests/test_imputer.py
Comment (dsherry) [inline evalml/tests/component_tests/test_imputer.py:307]: Why were nans here filled with `True` rather than `False`? Your example data had one `True` and two `False`. I thought the default categorical strategy of `most_frequent` would fill nans with `False`.

File: evalml/pipelines/components/transformers/imputers/imputer.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/imputers/imputer.py:72]: @angela97lin heads up these changes will cause conflicts for #1423 (update all components to accept woodwork datatables). I think they should be pretty straightforward to address--when you're resolving the conflicts you could call `dt.set_logical_types(...)` to do this transformation, or you could keep doing the transformation with pandas. Great that @christopherbunn is adding test coverage for this case, that'll help!

**Review by dsherry — APPROVED** [review]
@christopherbunn thank you, looks good! I left one question about the behavior of the imputer for boolean inputs, may be a bug there.

File: evalml/pipelines/components/transformers/imputers/imputer.py
Comment (dsherry) [reply evalml/pipelines/components/transformers/imputers/imputer.py:72]: Ah, #1423 was just merged!

@christopherbunn if you need a hand updating this, please ping @angela97lin or I

Comment (angela97lin) [reply evalml/pipelines/components/transformers/imputers/imputer.py:72]: Oo yup, just saw this. Yes yes, please reach out if you need help fixing the merge conflicts!

**Comment by christopherbunn** [comment]
@dsherry Thanks for the review! I looked at the test case a bit deeper and part of the reason why that passes is that np.nan actually gets casted to True when the data type is set to bool. As an example, this snippet below

```
    X_multi = pd.DataFrame({
        "bool with nan": pd.Series([True, np.nan, False, np.nan, False], dtype=bool),
        "bool no nan": pd.Series([False, False, False, False, True], dtype=bool),
    })
``` 
is actually equal to
```
    pd.DataFrame({
        "bool with nan": pd.Series([True, True, False, True, False], dtype=bool),
        "bool no nan": pd.Series([False, False, False, False, True], dtype=bool),
    })
```
Because of this, I'm not actually sure if its possible to impute a bool-only df like `X_multi` since it will always be filled with values that are either True or False. `X_multi` could exist with True, False, and np.nan values but only if the dtype is `Categorical` from the beginning. In that case, the existing code would already work.

I think that resolving this issue should instead focus on just catching this value error. The current HEAD of this PR will just return all-bool dfs as is without imputing it and I added a new case to test this behavior. Thoughts @angela97lin and @dsherry?

**Comment by angela97lin** [comment]
@christopherbunn Yup, I'd agree that resolving this issue to just catch the ValueError sounds good to me. It's a limitation or design choice made by pandas, and one I came across as well while implementing Woodwork stuff. Now that we have woodwork integration in place for components, users can use Woodwork structures, which use the new nullable pd.NA, and this would be converted to a category (or pandas "object") column, which you've mentioned already works. (Maybe this is a worthwhile test to add for the imputer specifically?)

**Comment by dsherry** [comment]
Discussed with @angela97lin @christopherbunn . Recommendations:
* For `SimpleImputer`, if we get this error, catch the error and return the original data
* Unit tests should cover both pandas as input and woodwork as input

File: evalml/pipelines/components/transformers/imputers/simple_imputer.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/imputers/simple_imputer.py:51]: [Suggested alternative:](https://pandas.pydata.org/pandas-docs/version/0.25.3/reference/api/pandas.api.types.is_bool_dtype.html) `pd.ptypes.is_bool_dtype(X).all()`

File: evalml/pipelines/components/transformers/imputers/simple_imputer.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/imputers/simple_imputer.py:77]: Since this is kinda confusing, perhaps add a comment saying "bool dtype doesn't support nans, and sklearn errors if all cols are bool, so return early"

File: evalml/tests/component_tests/test_imputer.py
Comment (dsherry) [inline evalml/tests/component_tests/test_imputer.py:297]: Since `Imputer` uses `SimpleImputer` under the hood, and you're putting the fix into `SimpleImputer`, I think the coverage of `SimpleImputer` is enough and you can delete this copy of the unit test.

File: evalml/tests/component_tests/test_simple_imputer.py
Comment (dsherry) [inline evalml/tests/component_tests/test_simple_imputer.py:126]: This test is great!

I think the first two cases (i.e. above this line) are fully covered by the last case, right? I suggest you delete the first two cases, above this line, and keep the last one below.

File: docs/source/release_notes.rst
Comment (dsherry) [inline docs/source/release_notes.rst:28]: Suggest "Fix SimpleImputer error which occurs when all features are bool type"

**Review by dsherry — APPROVED** [review]
I left a few suggestions which we should resolve before merging but LGTM!

**Review by freddyaboulton — APPROVED** [review]
@christopherbunn I think this is fantastic! Thanks for making this change. I think your unit test for the `Imputer`  shows that `SimpleImputer` and `Imputer` convert dtypes between input and output differently (even though `Imputer` calls `SimpleImputer` under the hood). 

I don't think that should block merge (maybe that behavior existed before this change) so maybe the best thing is to address @dsherry 's comments (and any others that come up) and file an issue to continue the discussion.

File: evalml/tests/component_tests/test_imputer.py
Comment (christopherbunn) [reply evalml/tests/component_tests/test_imputer.py:297]: Yep, my understanding is that while `Imputer` does use `SimpleImputer` under the hood, the output of `Imputer` is different as it recasts everything to the proper type. I don't see an issue with changing this in the future.

File: evalml/pipelines/components/transformers/imputers/simple_imputer.py
Comment (christopherbunn) [reply evalml/pipelines/components/transformers/imputers/simple_imputer.py:51]: It might not have the same behavior, as it checks the individual data type rather than the series/dataframe type. As such, this would return True for types that are able to be handled by the imputer just fine (such as category). 

image

File: evalml/tests/component_tests/test_simple_imputer.py
Comment (christopherbunn) [reply evalml/tests/component_tests/test_simple_imputer.py:126]: I'd argue for keeping all of the test cases. The first test case triggers the type conversion while the second one checks for proper boolean impution. The third one is more of an extension of the second one but it also combines a complete boolean column to make sure that the types are being processed properly.

File: evalml/tests/component_tests/test_imputer.py
Comment (dsherry) [inline evalml/tests/component_tests/test_imputer.py:334]: Unit tests discussed with @christopherbunn 
1. All cols are bool-type
`X = pd.DataFrame([True, True, False, True, True], dtype=bool)`
2. Imputation strategies work for bool type
`X = pd.DataFrame([True, np.nan, False, np.nan, True], dtype=object)`
`X = pd.DataFrame([True, np.nan, False, np.nan, False], dtype=object)`
3. Multi-type, with at least one bool typed col
```python
    X_multi_expected_arr = pd.DataFrame({
        "bool with nan": pd.Series([True, False, False, False, False], dtype=object),
        "bool no nan": pd.Series([False, False, False, False, True], dtype=bool),
    })
```

File: evalml/pipelines/components/transformers/imputers/simple_imputer.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/imputers/simple_imputer.py:77]: @christopherbunn should this comment be in `fit` too?

**Review by dsherry — APPROVED** [review]
@christopherbunn thanks for chasing this down!

**Pull request merged**
